### PR TITLE
alternative serde method for pandas

### PR DESCRIPTION
**Pull request by Koukyosyumei**
## Description

#5215
I'm Investigating alternative serde method than dicts.

### results 
(mean ± std. dev. of 7 runs, 20 loops each)

task|dict|Arrow IPC|Arrow IPC with compression|feather|pickle|Arrow Parquet|
 ---|---|   ---  |---          |    ---      |    ---     | ---   | 
send (pd.Series (10k rows))|2.56s| 6.47 ms ± 492 µs |7.09 ms ± 1.51 ms|6.76 ms ± 1.58 ms |8.33ms|11.3 ms ± 7.09 ms|
send&get (pd.Series (10k rows))|5.45s   | 10.9 ms ± 189 µs |11.9 ms ± 1.74 ms|10.7 ms ± 1.13 ms|  10.3ms | 16.1 ms ± 1.39 ms |
send (small pd.DataFrame (1k rows))   |8,72s  | 14 ms ± 396 µs |12.3 ms ± 532 µs|20.9 ms ± 6.58 ms | 13.5ms  | 23.2 ms ± 1.63 ms |
send&get (small pd.DataFrame (1k rows)) || 22.1 ms ± 596 µs |23.7 ms ± 4.99 ms|21.4 ms ± 3.32 ms |  21.2ms | 31.5 ms ± 6.47 ms|
send (big pd.DataFrame (55k rows, 15 columns))  || 1.58 s ± 97.2 ms|559 ms ± 46.4 ms|620 ms ± 34.4 ms| 1.68s  |622 ms ± 106 ms|
send&get (big pd.DataFrame (55k rows, 15 columns))   || 2.96 s ± 101 ms|913 ms ± 24.5 ms|971 ms ± 127 ms| 3.45s  |1.14 s ± 198 ms|

### notes
・Msgpack has been removed from pandas support
・Feather format is exactly the Arrow IPC file format
・pandas.to_feather, to_parquest, and to_pickle automatically compress the object
・pandas.to_feather doesn't support non-default index
・Feather does not support serializing a non-default index like multi-index for the index.
・Pyarrow.serialize is deprecated in pyarrow 2.0
・According to the arrow document, "reading Parquet files generally requires efficient yet relatively complex decoding, while reading Arrow IPC files does not involve any decoding because the on-disk representation is the same as the in-memory representation."
・The arrow document also says, "Parquet files are often much smaller than Arrow IPC files because of the columnar data compression strategies that Parquet uses. If your disk storage or network is slow, Parquet may be a better choice even for short-term storage or caching."
・To use h5py, users have to create an account for the HDF group and install HDF5. Thus, I think h5py is not user-friendly.

### conclusion
To sum up, I suggest using Arrow IPC with compression since it performs well for both small and big objects.

### reference
・http://arrow.apache.org/faq/

## Affected Dependencies
pyarrow

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/syft/lib/pandas/frame.py [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+    return PandasDataFrame_PB(dataframe=pa.serialize(table).to_buffer().to_pybytes())
 
 
 def proto2object(proto: PandasDataFrame_PB) -> pd.DataFrame:
-    dataframe_dict = Dict._proto2object(proto=proto.dataframe)
-    return pd.DataFrame.from_dict(dataframe_dict.upcast())
+    """Convert PandasDataFrame_PB to pd.DataFrame with pyarrow.
+
+    # noqa: DAR101
+    # noqa: DAR201
```
Comment (Param-29) [inline src/syft/lib/pandas/frame.py:27]: NIPK: 

I believe documentation guys will not encourage this, what you can do instead is 

```
    """Proto to object conversion using to return desired model.

    Args:
        proto: Serialized version of Dataframe, which will be used to reconstruct model.

    Returns:
        Re-constructed dataframe.
    """
```

File: src/syft/lib/pandas/frame.py [hunk 2]
```diff
@@ -1,26 +1,34 @@
...
 from ...proto.lib.pandas.frame_pb2 import PandasDataFrame as PandasDataFrame_PB
 
 
 def object2proto(obj: pd.DataFrame) -> PandasDataFrame_PB:
-    pd_dict = PrimitiveFactory.generate_primitive(value=obj.to_dict())
-    dict_proto = pd_dict._object2proto()
+    """Convert pd.DataFrame to PandasDataFrame_PB with pyarrow.
+
+    # noqa: DAR101
+    # noqa: DAR201
```
Comment (Param-29) [inline src/syft/lib/pandas/frame.py:16]: Same as above!

File: src/syft/lib/pandas/series.py [hunk 3]
```diff
@@ -1,26 +1,34 @@
...
 from ...proto.lib.pandas.series_pb2 import PandasSeries as PandasSeries_PB
 
 
 def object2proto(obj: pd.Series) -> PandasSeries_PB:
-    series_dict = PrimitiveFactory.generate_primitive(value=obj.to_dict())
-    dict_proto = series_dict._object2proto()
+    """Convert pd.Series to PandasSeries_PB with pyarrow.
+
+    # noqa: DAR101
+    # noqa: DAR201
```
Comment (Param-29) [inline src/syft/lib/pandas/series.py:16]: Same as above!

File: src/syft/lib/pandas/series.py [hunk 4]
```diff
@@ -1,26 +1,34 @@
...
+    return PandasSeries_PB(series=pa.serialize(table).to_buffer().to_pybytes())
 
 
 def proto2object(proto: PandasSeries_PB) -> pd.Series:
-    series_dict = Dict._proto2object(proto=proto.series)
-    return pd.Series(series_dict.upcast())
+    """Convert PandasSeries_PB to pd.Series with pyarrow.
+
+    # noqa: DAR101
+    # noqa: DAR201
```
Comment (Param-29) [inline src/syft/lib/pandas/series.py:27]: Same as above!

**Review by Param-29 — COMMENTED** [review]
I have added in a couple of comments, hope it helps ahead.

**Comment by tudorcebere** [comment]
Awesome work! Before jumping and implementing this, we might want to discuss how to add custom serde methods to certain types in a scalable way before coding it. Can't wait to see this one live!

**Comment by Koukyosyumei** [comment]
@tudorcebere 
Do you mean allowing users to choose the serde method?

**Comment by madhavajay** [comment]
@Koukyosyumei awesome work, I love the benchmarks! Looks like `arrow` for the win!

File: src/syft/lib/pandas/frame.py (hunk 1 above)
Comment (jmaunon) [reply src/syft/lib/pandas/frame.py:27]: Thansk @Param-29 !
That is the idea, not just ignoring "errors",  it is needed to addressed them

Comment (Koukyosyumei) [reply src/syft/lib/pandas/frame.py]: Thanks for your review!
